File: php/core/Context.php
<?php
declare(strict_types=1);

// IbanValidation SDK context

require_once __DIR__ . '/Control.php';
require_once __DIR__ . '/Operation.php';
require_once __DIR__ . '/Spec.php';
require_once __DIR__ . '/Result.php';
require_once __DIR__ . '/Response.php';
require_once __DIR__ . '/Error.php';
require_once __DIR__ . '/Helpers.php';

class IbanValidationContext
{
    public string $id;
    public array $out;
    public mixed $client;
    public ?IbanValidationUtility $utility;
    public IbanValidationControl $ctrl;
    public array $meta;
    public ?array $config;
    public ?array $entopts;
    public ?array $options;
    public mixed $entity;
    public ?array $shared;
    public array $opmap;
    public array $data;
    public array $reqdata;
    public array $match;
    public array $reqmatch;
    public ?array $point;
    public ?IbanValidationSpec $spec;
    public ?IbanValidationResult $result;
    public ?IbanValidationResponse $response;
    public IbanValidationOperation $op;

    public function __construct(array $ctxmap = [], ?self $basectx = null)
    {
        $this->id = 'C' . random_int(10000000, 99999999);
        $this->out = [];

        $this->client = IbanValidationHelpers::get_ctx_prop($ctxmap, 'client') ?? ($basectx ? $basectx->client : null);
        $this->utility = IbanValidationHelpers::get_ctx_prop($ctxmap, 'utility') ?? ($basectx ? $basectx->utility : null);

        $this->ctrl = new IbanValidationControl();
        $ctrl_raw = IbanValidationHelpers::get_ctx_prop($ctxmap, 'ctrl');
        if (is_array($ctrl_raw)) {
            if (array_key_exists('throw', $ctrl_raw)) {
                $this->ctrl->throw_err = $ctrl_raw['throw'];
            }
            if (isset($ctrl_raw['explain']) && is_array($ctrl_raw['explain'])) {
                $this->ctrl->explain = $ctrl_raw['explain'];
            }
        } elseif ($basectx !== null && $basectx->ctrl !== null) {
            $this->ctrl = $basectx->ctrl;
        }

        $m = IbanValidationHelpers::get_ctx_prop($ctxmap, 'meta');
        $this->meta = is_array($m) ? $m : ($basectx ? $basectx->meta ?? [] : []);

        $cfg = IbanValidationHelpers::get_ctx_prop($ctxmap, 'config');
        $this->config = is_array($cfg) ? $cfg : ($basectx ? $basectx->config : null);

        $eo = IbanValidationHelpers::get_ctx_prop($ctxmap, 'entopts');
        $this->entopts = is_array($eo) ? $eo : ($basectx ? $basectx->entopts : null);

        $o = IbanValidationHelpers::get_ctx_prop($ctxmap, 'options');
        $this->options = is_array($o) ? $o : ($basectx ? $basectx->options : null);

        $e = IbanValidationHelpers::get_ctx_prop($ctxmap, 'entity');
        $this->entity = $e ?? ($basectx ? $basectx->entity : null);

        $s = IbanValidationHelpers::get_ctx_prop($ctxmap, 'shared');
        $this->shared = is_array($s) ? $s : ($basectx ? $basectx->shared : null);

        $om = IbanValidationHelpers::get_ctx_prop($ctxmap, 'opmap');
        $this->opmap = is_array($om) ? $om : ($basectx ? $basectx->opmap ?? [] : []);

        $this->data = IbanValidationHelpers::to_map(IbanValidationHelpers::get_ctx_prop($ctxmap, 'data')) ?? [];
        $this->reqdata = IbanValidationHelpers::to_map(IbanValidationHelpers::get_ctx_prop($ctxmap, 'reqdata')) ?? [];
        $this->match = IbanValidationHelpers::to_map(IbanValidationHelpers::get_ctx_prop($ctxmap, 'match')) ?? [];
        $this->reqmatch = IbanValidationHelpers::to_map(IbanValidationHelpers::get_ctx_prop($ctxmap, 'reqmatch')) ?? [];

        $pt = IbanValidationHelpers::get_ctx_prop($ctxmap, 'point');
        $this->point = is_array($pt) ? $pt : ($basectx ? $basectx->point : null);

        $sp = IbanValidationHelpers::get_ctx_prop($ctxmap, 'spec');
        $this->spec = ($sp instanceof IbanValidationSpec) ? $sp : ($basectx ? $basectx->spec : null);

        $r = IbanValidationHelpers::get_ctx_prop($ctxmap, 'result');
        $this->result = ($r instanceof IbanValidationResult) ? $r : ($basectx ? $basectx->result : null);

        $rp = IbanValidationHelpers::get_ctx_prop($ctxmap, 'response');
        $this->response = ($rp instanceof IbanValidationResponse) ? $rp : ($basectx ? $basectx->response : null);

        $opname = IbanValidationHelpers::get_ctx_prop($ctxmap, 'opname') ?? '';
        $this->op = $this->resolve_op($opname);
    }

    public function resolve_op(string $opname): IbanValidationOperation
    {
        // Cache key is `<entity>:<opname>` so two entities with the same op
        // (e.g. both have a "list") get distinct cached Operations. Keying
        // on opname alone caused the first-resolved entity's points to be
        // served to every subsequent entity's call.
        $entname = (is_object($this->entity) && method_exists($this->entity, 'get_name'))
            ? $this->entity->get_name()
            : '_';
        $cacheKey = $entname . ':' . $opname;

        if (isset($this->opmap[$cacheKey])) {
            return $this->opmap[$cacheKey];
        }
        if ($opname === '') {
            return new IbanValidationOperation([]);
        }

        $opcfg = \Voxgig\Struct\Struct::getpath($this->config, "entity.{$entname}.op.{$opname}");

        $input = ($opname === 'update' || $opname === 'create') ? 'data' : 'match';

        $points = [];
        if (is_array($opcfg)) {
            $t = \Voxgig\Struct\Struct::getprop($opcfg, 'points');
            if (is_array($t)) {
                $points = $t;
            }
        }

        $op = new IbanValidationOperation([
            'entity' => $entname,
            'name' => $opname,
            'input' => $input,
            'points' => $points,
        ]);
        $this->opmap[$cacheKey] = $op;
        return $op;
    }

    public function make_error(string $code, string $msg): IbanValidationError
    {
        return new IbanValidationError($code, $msg, $this);
    }
}

File: .sdk/tm/rb/core/context.rb
# IbanValidation SDK context

require_relative '../utility/struct/voxgig_struct'
require_relative 'control'
require_relative 'operation'
require_relative 'spec'
require_relative 'result'
require_relative 'response'
require_relative 'error'
require_relative 'helpers'

class IbanValidationContext
  attr_accessor :id, :out, :client, :utility, :ctrl, :meta, :config,
                :entopts, :options, :entity, :shared, :opmap,
                :data, :reqdata, :match, :reqmatch, :point,
                :spec, :result, :response, :op

  def initialize(ctxmap = {}, basectx = nil)
    ctxmap ||= {}
    @id = "C#{rand(10000000..99999999)}"
    @out = {}

    @client = IbanValidationHelpers.get_ctx_prop(ctxmap, "client") || basectx&.client
    @utility = IbanValidationHelpers.get_ctx_prop(ctxmap, "utility") || basectx&.utility

    @ctrl = IbanValidationControl.new
    ctrl_raw = IbanValidationHelpers.get_ctx_prop(ctxmap, "ctrl")
    if ctrl_raw.is_a?(Hash)
      @ctrl.throw_err = ctrl_raw["throw"] if ctrl_raw.key?("throw")
      @ctrl.explain = ctrl_raw["explain"] if ctrl_raw["explain"].is_a?(Hash)
      @ctrl.actor = ctrl_raw["actor"] if ctrl_raw.key?("actor")
      @ctrl.paging = ctrl_raw["paging"] if ctrl_raw["paging"].is_a?(Hash)
    elsif basectx&.ctrl
      @ctrl = basectx.ctrl
    end

    m = IbanValidationHelpers.get_ctx_prop(ctxmap, "meta")
    @meta = m.is_a?(Hash) ? m : (basectx&.meta || {})

    cfg = IbanValidationHelpers.get_ctx_prop(ctxmap, "config")
    @config = cfg.is_a?(Hash) ? cfg : basectx&.config

    eo = IbanValidationHelpers.get_ctx_prop(ctxmap, "entopts")
    @entopts = eo.is_a?(Hash) ? eo : basectx&.entopts

    o = IbanValidationHelpers.get_ctx_prop(ctxmap, "options")
    @options = o.is_a?(Hash) ? o : basectx&.options

    e = IbanValidationHelpers.get_ctx_prop(ctxmap, "entity")
    @entity = e || basectx&.entity

    s = IbanValidationHelpers.get_ctx_prop(ctxmap, "shared")
    @shared = s.is_a?(Hash) ? s : basectx&.shared

    om = IbanValidationHelpers.get_ctx_prop(ctxmap, "opmap")
    @opmap = om.is_a?(Hash) ? om : (basectx&.opmap || {})

    @data = IbanValidationHelpers.to_map(IbanValidationHelpers.get_ctx_prop(ctxmap, "data")) || {}
    @reqdata = IbanValidationHelpers.to_map(IbanValidationHelpers.get_ctx_prop(ctxmap, "reqdata")) || {}
    @match = IbanValidationHelpers.to_map(IbanValidationHelpers.get_ctx_prop(ctxmap, "match")) || {}
    @reqmatch = IbanValidationHelpers.to_map(IbanValidationHelpers.get_ctx_prop(ctxmap, "reqmatch")) || {}

    pt = IbanValidationHelpers.get_ctx_prop(ctxmap, "point")
    @point = pt.is_a?(Hash) ? pt : basectx&.point

    sp = IbanValidationHelpers.get_ctx_prop(ctxmap, "spec")
    @spec = sp.is_a?(IbanValidationSpec) ? sp : basectx&.spec

    r = IbanValidationHelpers.get_ctx_prop(ctxmap, "result")
    @result = r.is_a?(IbanValidationResult) ? r : basectx&.result

    rp = IbanValidationHelpers.get_ctx_prop(ctxmap, "response")
    @response = rp.is_a?(IbanValidationResponse) ? rp : basectx&.response

    opname = IbanValidationHelpers.get_ctx_prop(ctxmap, "opname") || ""
    @op = resolve_op(opname)
  end

  def resolve_op(opname)
    # Cache key is `<entity>:<opname>` so two entities with the same op
    # (e.g. both have a "list") get distinct cached Operations. Keying
    # on opname alone caused the first-resolved entity's points to be
    # served to every subsequent entity's call.
    entname = @entity&.respond_to?(:get_name) ? @entity.get_name : "_"
    cache_key = "#{entname}:#{opname}"
    return @opmap[cache_key] if @opmap[cache_key]
    return IbanValidationOperation.new({}) if opname.empty?

    opcfg = VoxgigStruct.getpath(@config, "entity.#{entname}.op.#{opname}")

    input = (opname == "update" || opname == "create") ? "data" : "match"

    points = []
    if opcfg.is_a?(Hash)
      t = VoxgigStruct.getprop(opcfg, "points")
      points = t if t.is_a?(Array)
    end

    op = IbanValidationOperation.new({
      "entity" => entname,
      "name" => opname,
      "input" => input,
      "points" => points,
    })
    @opmap[cache_key] = op
    op
  end

  def make_error(code, msg)
    IbanValidationError.new(code, msg, self)
  end
end
